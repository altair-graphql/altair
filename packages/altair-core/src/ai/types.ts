export interface ISession {
  id: string;
  userId: string;
  title: string;
  isActive: boolean;
}

export interface IMessage {
  id: string;
  sessionId: string;
  message: string;
  role: 'USER' | 'ASSISTANT';
}

export interface ISendMessageDto {
  message: string;
  sdl?: string;
  graphqlQuery?: string;
  graphqlVariables?: string;
}

export interface IRateMessageDto {
  rating: number;
}

