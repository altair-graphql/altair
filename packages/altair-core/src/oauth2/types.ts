export enum OAuth2Type {
  AUTHORIZATION_CODE = 'auth_code',
  AUTHORIZATION_CODE_PKCE = 'auth_code_pkce',
}

export interface SimpleAuthorizationRequestParams extends Record<string, string> {
  response_type: 'code';
  /**
   * The client ID of your application
   */
  client_id: string;
  /**
   * The URL in your app where users will be sent after authorization
   */
  redirect_uri: string;
  /**
   * A random string generated by your application, which you'll verify later
   */
  state: string;
  /**
   * A space-separated list of scopes
   */
  scope: string;
}
export interface CodeChallengeAuthorizationRequestParams
  extends SimpleAuthorizationRequestParams {
  code_challenge: string;
  code_challenge_method: 'S256';
}
export type AuthorizationRequestParams =
  | SimpleAuthorizationRequestParams
  | CodeChallengeAuthorizationRequestParams;

export interface AuthorizationRedirectResponse {
  /**
   * The authorization code received from the authorization server
   */
  code: string;
  /**
   * The exact value of the state parameter passed by the client in the authorization request
   */
  state: string;
}

export interface AuthorizationRedirectErrorResponse {
  error: string;
  state: string;
  error_description?: string;
  error_uri?: string;
}

export interface AuthorizationCode_AccessTokenRequest {
  grant_type: 'authorization_code';
  /**
   * The authorization code received from the authorization server
   */
  code: string;
  /**
   * The redirect URI used in the initial request
   */
  redirect_uri: string;
  client_id: string;
  client_secret: string;
}
export interface AuthorizationCodePKCE_AccessTokenRequest
  extends AuthorizationCode_AccessTokenRequest {
  code_verifier: string;
}
export type AccessTokenRequest =
  | AuthorizationCode_AccessTokenRequest
  | AuthorizationCodePKCE_AccessTokenRequest;

export interface AccessTokenResponse {
  /**
   * The access token issued by the authorization server
   */
  access_token: string;
  /**
   * The type of the token issued
   */
  token_type: string;
  /**
   * The exact value of the state parameter passed by the client in the authorization request
   */
  state: string;
  /**
   * The lifetime in seconds of the access token
   */
  expires_in?: number;
  refresh_token?: string;
  id_token?: string;
  scope?: string;
}

export interface AccessTokenErrorResponse {
  error: string;
  state: string;
  error_description?: string;
  error_uri?: string;
}
