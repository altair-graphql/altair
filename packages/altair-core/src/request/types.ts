import { Observable } from 'rxjs';
import { HeaderState } from '../types/state/header.interfaces';
import { SelectedOperation } from '../types/state/query.interfaces';

interface ResolvedFileVariable {
  name: string;
  data: File;
}
export interface GraphQLRequestOptions {
  url: string;
  query: string;
  method: string;
  withCredentials?: boolean;
  variables?: Record<string, unknown>;
  extensions?: Record<string, unknown>;
  headers?: HeaderState;
  files?: ResolvedFileVariable[];
  selectedOperation?: SelectedOperation;
  batchedRequest?: boolean;
  additionalParams?: Record<string, unknown>;
}
export interface GraphQLResponseData {
  ok: boolean;
  data: string;
  headers: Headers;
  status: number;
  statusText: string;
  url: string;

  /**
   * The timestamp when the request was started
   */
  requestStartTimestamp: number;

  /**
   * The timestamp when the request was ended
   */
  requestEndTimestamp: number;

  /**
   * The time taken to get the response in milliseconds
   */
  resopnseTimeMs: number;
}

export interface GraphQLRequestHandler {
  handle(request: GraphQLRequestOptions): Observable<GraphQLResponseData>;
  generateCurl(request: GraphQLRequestOptions): string;
  destroy?(): void;
}

export enum MultiResponseStrategy {
  /**
   * Automatically determine the strategy based on the response
   */
  AUTO = 'auto',

  /**
   * Concatenate all responses
   */
  CONCATENATE = 'concatenate',

  /**
   * Append responses as a list
   */
  APPEND = 'append',

  /**
   * Patch the responses together following the GraphQL spec
   */
  PATCH = 'patch',
}

export const HTTP_HANDLER_ID = 'http';
export const WEBSOCKET_HANDLER_ID = 'websocket';
export const GRAPHQL_WS_HANDLER_ID = 'graphql-ws';
export const APP_SYNC_HANDLER_ID = 'app-sync';
export const ACTION_CABLE_HANDLER_ID = 'action-cable';
export const GRAPHQL_SSE_HANDLER_ID = 'graphql-sse';

export const REQUEST_HANDLER_IDS = {
  HTTP: HTTP_HANDLER_ID,
  WEBSOCKET: WEBSOCKET_HANDLER_ID,
  GRAPHQL_WS: GRAPHQL_WS_HANDLER_ID,
  APP_SYNC: APP_SYNC_HANDLER_ID,
  ACTION_CABLE: ACTION_CABLE_HANDLER_ID,
  GRAPHQL_SSE: GRAPHQL_SSE_HANDLER_ID,
} as const;

export type RequestHandlerIds =
  (typeof REQUEST_HANDLER_IDS)[keyof typeof REQUEST_HANDLER_IDS];

export interface RequestHandlerData {
  /**
   * Unique identifier for the handler
   */
  id: string;

  /**
   * Function that returns a promise that resolves with the handler class (NOT an instance of the class)
   */
  getHandler: () => Promise<GraphQLRequestHandler>;

  /**
   * The text to be shown for this handler in the Altair UI
   */
  copyTag?: string;
}
