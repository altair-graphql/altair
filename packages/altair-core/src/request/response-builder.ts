import { GraphQLError } from 'graphql';
import { parseJson } from '../utils/json';
import { MultiResponseStrategy } from './types';
import { setByDotNotation } from '../utils/dot-notation';
import { QueryResponse } from '../types/state/query.interfaces';
type ErrorLike = Partial<GraphQLError> | Error;
type Extensions = Record<string, any>;
type Path = (string | number)[];

// Snapshot: https://github.com/graphql/graphql-spec/blob/c630301560d9819d33255d3ba00f548e8abbcdc6/spec/Section%207%20--%20Response.md#incremental
interface IncrementalStreamResult {
  items: unknown[] | null;
  path: Path;
  label?: string;
  errors?: ErrorLike[];
  extensions?: Extensions;
}

interface IncrementalDeferResult {
  data: Record<string, unknown> | null;
  path: Path;
  label?: string;
  errors?: ErrorLike[];
  extensions?: Extensions;
}

type IncrementalResult = IncrementalStreamResult | IncrementalDeferResult;

interface InitialExecutionResult {
  data?: null | Record<string, any>;
  errors?: ErrorLike[];
  extensions?: Extensions;
  hasNext?: boolean;
}

interface OldSubsequentExecutionResult extends InitialExecutionResult {
  path: Path;
}
interface SubsequentExecutionResultWithIncremental {
  incremental?: IncrementalResult[];
  errors?: ErrorLike[];
  extensions?: Extensions;
  hasNext: true;
}
interface SubsequentExecutionResultFinal {
  errors?: ErrorLike[];
  extensions?: Extensions;
  hasNext: false;
}

type SubsequentExecutionResult =
  | SubsequentExecutionResultWithIncremental
  | OldSubsequentExecutionResult
  | SubsequentExecutionResultFinal;

type ExecutionResult = InitialExecutionResult | SubsequentExecutionResult;

// https://github.com/graphql/graphql-wg/blob/main/rfcs/DeferStream.md
// https://github.com/graphql/graphql-over-http/blob/main/rfcs/IncrementalDelivery.md
// https://github.com/graphql/graphql-spec/blob/c630301560d9819d33255d3ba00f548e8abbcdc6/spec/Section%207%20--%20Response.md#incremental
// https://github.com/graphql/defer-stream-wg/discussions/69
export const buildResponse = (
  responses: QueryResponse[],
  strategy = MultiResponseStrategy.AUTO
): QueryResponse[] => {
  switch (strategy) {
    case MultiResponseStrategy.CONCATENATE:
      return buildResponse__concatenate(responses);
    case MultiResponseStrategy.APPEND:
      return buildResponse__append(responses);
    case MultiResponseStrategy.PATCH:
      return buildResponse__patch(responses);
    case MultiResponseStrategy.AUTO: {
      const firstResponse = responses[0];
      const parsedFirstResponse = parseJson(firstResponse?.content ?? '', {
        defaultValue: null,
      });
      // if response[0] is not a JSON object, then concatenate
      if (!parsedFirstResponse || typeof parsedFirstResponse !== 'object') {
        return buildResponse__concatenate(responses);
      }
      // if response[0] is patchable, then patch
      if ('data' in parsedFirstResponse && 'hasNext' in parsedFirstResponse) {
        return buildResponse__patch(responses);
      }
      // otherwise append
      return buildResponse__append(responses);
    }
    default:
      throw new Error('Invalid strategy');
  }
};

const buildResponse__concatenate = (responses: QueryResponse[]): QueryResponse[] => {
  const content = responses.map((r) => r.content).join('');
  const parsedContent = parseJson(content, {defaultValue: null});
  return [
    {
      content: parsedContent ? JSON.stringify(parsedContent, null, 2) : content,
      timestamp: responses.at(-1)?.timestamp ?? 0,
    },
  ];
};

const buildResponse__append = (responses: QueryResponse[]): QueryResponse[] => {
  return responses;
};

const buildResponse__patch = (responses: QueryResponse[]): QueryResponse[] => {
  if (responses.length === 0) {
    return [];
  }
  // first response is the same shape as a standard GraphQL response
  const obj = parseJson<InitialExecutionResult, null>(responses[0]?.content ?? '', {
    defaultValue: null,
  });

  if (!obj) {
    throw new Error('JSON response required for patching!');
  }
  // remove hasNext field from obj (since it's just used for patching)
  Reflect.deleteProperty(obj, 'hasNext');

  for (let i = 1; i < responses.length; i++) {
    const nextObj = parseJson<SubsequentExecutionResult, null>(
      responses[i]?.content ?? '',
      { defaultValue: null }
    );

    if (!nextObj) {
      throw new Error('JSON response required for patching!');
    }
    patchResponse(obj, nextObj);
  }
  // TODO: Handle cases with labels

  // return the patched response
  return [
    {
      content: JSON.stringify(obj, null, 2),
      timestamp: responses.at(0)?.timestamp ?? 0,
    },
  ];
};

const patchResponse = (
  obj: InitialExecutionResult,
  nextData: SubsequentExecutionResult
): ExecutionResult => {
  const result = { ...obj };
  let errors = result.errors ? result.errors : [];

  const extensions = {
    ...result.extensions,
    ...('extensions' in nextData ? nextData.extensions : {}),
  };

  if (nextData.errors) {
    errors.push(...nextData.errors);
  }

  if ('path' in nextData) {
    // old version of the incremental delivery payloads
    // https://github.com/graphql/graphql-over-http/blob/d80afd45782b40a9a8447fcff3d772689d83df56/rfcs/IncrementalDelivery.md
    nextData = {
      ...nextData,
      incremental: [
        {
          data: nextData.data ?? null,
          path: nextData.path,
        },
      ],
      hasNext: true,
    } satisfies SubsequentExecutionResultWithIncremental;
  }

  if (nextData.hasNext && nextData.incremental) {
    for (const patch of nextData.incremental) {
      if (Array.isArray(patch.errors)) {
        errors.push(...patch.errors);
      }

      if (patch.extensions) {
        Object.assign(extensions, patch.extensions);
      }

      const path = ['data', ...patch.path];
      if ('data' in patch) {
        // @defer patch
        // set the data at the path in result
        if (patch.data) {
          result.data = result.data ?? {};
          setByDotNotation(result, path, patch.data, true);
        }
      }
      if ('items' in patch) {
        // @stream patch
        patch.items?.forEach((item, i) => {
          const lastPathIdx = path[path.length - 1];
          if (typeof lastPathIdx !== 'number') {
            throw new Error(
              'Path for stream incremental payload should end with a number!'
            );
          }
          // for each item, we want to place it at the end of the array
          const newPath = [...path.slice(0, -1), lastPathIdx + i];
          setByDotNotation(result, newPath, item);
        });
      }
    }
  }

  return {
    ...result,
    ...(errors.length && { errors }),
    ...(Object.keys(extensions).length && { extensions }),
  };
};
