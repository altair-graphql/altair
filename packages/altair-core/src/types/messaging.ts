import { ExportWindowState } from './state/window.interfaces';

type WebExtensionPingMessage = {
  type: 'ping';
};

type WebExtensionPongMessage = {
  type: 'pong';
};

type WebExtensionReadyMessage = {
  type: 'ready';
};

type WebExtensionImportWindowMessage = {
  type: 'import-window';
  window: ExportWindowState;
};

export type WebExtensionMessage =
  | WebExtensionImportWindowMessage
  | WebExtensionPingMessage
  | WebExtensionPongMessage
  | WebExtensionReadyMessage;
