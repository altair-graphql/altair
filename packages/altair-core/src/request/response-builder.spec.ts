import { describe, expect, it } from '@jest/globals';
import { buildResponse } from './response-builder';
import { QueryResponse } from '../types/state/query.interfaces';
import { MultiResponseStrategy } from './types';

const responseChunks: QueryResponse[] = [
  {
    content:
      '{"data":{"hello":"Hello world","alphabet":[],"fastField":"This field resolves fast! ⚡️"},"hasNext":true}',
    timestamp: 1718252802585,
  },
  {
    content:
      '{"incremental":[{"items":["a"],"path":["alphabet",0]}],"hasNext":true}',
    timestamp: 1718252802585,
  },
  {
    content:
      '{"incremental":[{"items":["b"],"path":["alphabet",1]}],"hasNext":true}',
    timestamp: 1718252802585,
  },
  {
    content:
      '{"incremental":[{"items":["c"],"path":["alphabet",2]}],"hasNext":true}',
    timestamp: 1718252802585,
  },
  {
    content:
      '{"incremental":[{"items":["d"],"path":["alphabet",3]}],"hasNext":true}',
    timestamp: 1718252802585,
  },
  {
    content:
      '{"incremental":[{"items":["e"],"path":["alphabet",4]}],"hasNext":true}',
    timestamp: 1718252802585,
  },
  {
    content:
      '{"incremental":[{"data":{"slowField":"This field resolves slowly after 5000ms ⏳"},"path":[]}],"hasNext":true}',
    timestamp: 1718252802585,
  },
  {
    content:
      '{"incremental":[{"items":["f"],"path":["alphabet",5]}],"hasNext":true}',
    timestamp: 1718252802585,
  },
  {
    content:
      '{"incremental":[{"items":["g"],"path":["alphabet",6]}],"hasNext":true}',
    timestamp: 1718252802585,
  },
  {
    content:
      '{"incremental":[{"items":["h"],"path":["alphabet",7]}],"hasNext":true}',
    timestamp: 1718252802585,
  },
  {
    content:
      '{"incremental":[{"items":["i"],"path":["alphabet",8]}],"hasNext":true}',
    timestamp: 1718252802585,
  },
  {
    content:
      '{"incremental":[{"items":["j"],"path":["alphabet",9]}],"hasNext":true}',
    timestamp: 1718252802585,
  },
  {
    content:
      '{"incremental":[{"items":["k"],"path":["alphabet",10]}],"hasNext":true}',
    timestamp: 1718252802585,
  },
  {
    content:
      '{"incremental":[{"items":["l"],"path":["alphabet",11]}],"hasNext":true}',
    timestamp: 1718252802585,
  },
  {
    content:
      '{"incremental":[{"items":["m"],"path":["alphabet",12]}],"hasNext":true}',
    timestamp: 1718252802585,
  },
  {
    content:
      '{"incremental":[{"items":["n"],"path":["alphabet",13]}],"hasNext":true}',
    timestamp: 1718252802585,
  },
  {
    content:
      '{"incremental":[{"items":["o"],"path":["alphabet",14]}],"hasNext":true}',
    timestamp: 1718252802585,
  },
  {
    content:
      '{"incremental":[{"items":["p"],"path":["alphabet",15]}],"hasNext":true}',
    timestamp: 1718252802585,
  },
  {
    content:
      '{"incremental":[{"items":["q"],"path":["alphabet",16]}],"hasNext":true}',
    timestamp: 1718252802585,
  },
  {
    content:
      '{"incremental":[{"items":["r"],"path":["alphabet",17]}],"hasNext":true}',
    timestamp: 1718252802585,
  },
  {
    content:
      '{"incremental":[{"items":["s"],"path":["alphabet",18]}],"hasNext":true}',
    timestamp: 1718252802585,
  },
  {
    content:
      '{"incremental":[{"items":["t"],"path":["alphabet",19]}],"hasNext":true}',
    timestamp: 1718252802585,
  },
  {
    content:
      '{"incremental":[{"items":["u"],"path":["alphabet",20]}],"hasNext":true}',
    timestamp: 1718252802585,
  },
  {
    content:
      '{"incremental":[{"items":["v"],"path":["alphabet",21]}],"hasNext":true}',
    timestamp: 1718252802585,
  },
  {
    content:
      '{"incremental":[{"items":["w"],"path":["alphabet",22]}],"hasNext":true}',
    timestamp: 1718252802585,
  },
  {
    content:
      '{"incremental":[{"items":["x"],"path":["alphabet",23]}],"hasNext":true}',
    timestamp: 1718252802585,
  },
  {
    content:
      '{"incremental":[{"items":["y"],"path":["alphabet",24]}],"hasNext":true}',
    timestamp: 1718252802585,
  },
  {
    content:
      '{"incremental":[{"items":["z"],"path":["alphabet",25]}],"hasNext":true}',
    timestamp: 1718252802585,
  },
  {
    content: '{"hasNext":false}',
    timestamp: 1718252802585,
  },
];
describe('response-builder', () => {
  it('unknown strategy should throw an error', () => {
    expect(() =>
      buildResponse(responseChunks, 'unknown' as MultiResponseStrategy)
    ).toThrowError('Invalid strategy');
  });
  describe('concatenate strategy', () => {
    it('should concatenate the responses', () => {
      const res = buildResponse(responseChunks, MultiResponseStrategy.CONCATENATE);
      expect(res).toMatchSnapshot();
    });

    it('should return timestamp as 0 if no responses are provided', () => {
      const res = buildResponse([], MultiResponseStrategy.CONCATENATE);
      expect(res).toEqual([
        {
          content: '',
          timestamp: 0,
        },
      ]);
    });
  });

  describe('append strategy', () => {
    it('should append the responses', () => {
      const res = buildResponse(responseChunks, MultiResponseStrategy.APPEND);
      expect(res).toMatchSnapshot();
    });
  });

  describe('patch strategy', () => {
    it('should patch the responses', () => {
      const res = buildResponse(responseChunks, MultiResponseStrategy.PATCH);
      expect(res).toMatchSnapshot();
    });

    it('should return empty list if no responses are provided', () => {
      const res = buildResponse([], MultiResponseStrategy.PATCH);
      expect(res).toEqual([]);
    });

    it('should throw an error if the first response is not JSON', () => {
      expect(() =>
        buildResponse(
          [
            {
              content: 'This is not a JSON object',
              timestamp: 1718252802585,
            },
            {
              content:
                '{"incremental":[{"items":["a"],"path":["alphabet",0]}],"hasNext":true}',
              timestamp: 1718252802585,
            },
          ],
          MultiResponseStrategy.PATCH
        )
      ).toThrowError('JSON response required for patching!');
    });

    it('should throw an error if subsequent responses are not JSON', () => {
      expect(() =>
        buildResponse(
          [
            {
              content:
                '{"data":{"hello":"Hello world","alphabet":[],"fastField":"This field resolves fast! ⚡️"},"hasNext":true}',
              timestamp: 1718252802585,
            },
            {
              content: 'This is not a JSON object',
              timestamp: 1718252802585,
            },
          ],
          MultiResponseStrategy.PATCH
        )
      ).toThrowError('JSON response required for patching!');
    });

    it('should gather all errors from all responses', () => {
      const res = buildResponse(
        [
          {
            content:
              '{"data":{"hello":"Hello world","alphabet":[],"fastField":"This field resolves fast! ⚡️"},"hasNext":true,"errors":[{"message":"Error 1"}]}',
            timestamp: 1718252802585,
          },
          {
            content:
              '{"incremental":[{"items":["a"],"path":["alphabet",0],"errors":[{"message":"Incremental error"}]}],"hasNext":true,"errors":[{"message":"Error 2"}]}',
            timestamp: 1718252802585,
          },
          {
            content:
              '{"incremental":[{"items":["b"],"path":["alphabet",1]}],"hasNext":true,"errors":[{"message":"Error 3"}]}',
            timestamp: 1718252802585,
          },
        ],
        MultiResponseStrategy.PATCH
      );

      expect(res).toMatchSnapshot();
    });

    it('should handle incremental payloads in the old format', () => {
      const res = buildResponse(
        [
          {
            content:
              '{"data":{"hello":"Hello world","alphabet":[],"fastField":"This field resolves fast! ⚡️"},"hasNext":true}',
            timestamp: 1718252802585,
          },
          {
            content:
              '{"data":{"additionalField":"This field is added in the second response"},"path":["inner"]}',
            timestamp: 1718252802585,
          },
        ],
        MultiResponseStrategy.PATCH
      );

      expect(res).toMatchSnapshot();
    });

    it('should patch responses with extensions', () => {
      const res = buildResponse(
        [
          {
            content:
              '{"data":{"hello":"Hello world","alphabet":[],"fastField":"This field resolves fast! ⚡️"},"hasNext":true}',
            timestamp: 1718252802585,
          },
          {
            content:
              '{"incremental":[{"items":["a"],"path":["alphabet",0]}],"hasNext":true,"extensions":{"key":"value"}}',
            timestamp: 1718252802585,
          },
          {
            content:
              '{"incremental":[{"items":["b"],"path":["alphabet",1],"extensions":{"key1":"value1"}}],"hasNext":true}',
            timestamp: 1718252802585,
          },
        ],
        MultiResponseStrategy.PATCH
      );

      expect(res).toMatchSnapshot();
    });

    it('should throw an error for incremental payload with invalid @stream path', () => {
      expect(() =>
        buildResponse([
          {
            content:
              '{"data":{"hello":"Hello world","alphabet":[],"fastField":"This field resolves fast! ⚡️"},"hasNext":true}',
            timestamp: 1718252802585,
          },
          {
            content:
              '{"incremental":[{"items":["a"],"path":["alphabet"]}],"hasNext":true}',
            timestamp: 1718252802585,
          },
        ])
      ).toThrowError(
        'Path for stream incremental payload should end with a number!'
      );
    });
  });

  describe('auto strategy', () => {
    it('should concatenate the responses when the first response is not a JSON object', () => {
      const res = buildResponse(
        [
          {
            content: 'This is not a JSON object',
            timestamp: 1718252802585,
          },
          {
            content:
              '{"incremental":[{"items":["a"],"path":["alphabet",0]}],"hasNext":true}',
            timestamp: 1718252802585,
          },
        ],
        MultiResponseStrategy.AUTO
      );
      expect(res).toMatchSnapshot();
    });

    it('should return timestamp as 0 if no responses are provided', () => {
      const res = buildResponse([], MultiResponseStrategy.AUTO);
      expect(res).toEqual([
        {
          content: '',
          timestamp: 0,
        },
      ]);
    });

    it('should patch the responses when the first response is patchable', () => {
      const res = buildResponse(
        [
          {
            content:
              '{"data":{"hello":"Hello world","alphabet":[],"fastField":"This field resolves fast! ⚡️"},"hasNext":true}',
            timestamp: 1718252802585,
          },
          {
            content:
              '{"incremental":[{"items":["a"],"path":["alphabet",0]}],"hasNext":true}',
            timestamp: 1718252802585,
          },
        ],
        MultiResponseStrategy.AUTO
      );
      expect(res).toMatchSnapshot();
    });

    it('should append the responses when the first response is a JSON object but not patchable', () => {
      const res = buildResponse(
        [
          {
            content:
              '{"hello":"Hello world","alphabet":[],"fastField":"This field resolves fast! ⚡️"}',
            timestamp: 1718252802585,
          },
          {
            content: '{"hi": "Hi world"}',
            timestamp: 1718252802585,
          },
        ],
        MultiResponseStrategy.AUTO
      );
      expect(res).toMatchSnapshot();
    });
  });
});
