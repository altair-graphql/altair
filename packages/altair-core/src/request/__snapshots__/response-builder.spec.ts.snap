// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`response-builder append strategy should append the responses 1`] = `
[
  {
    "content": "{"data":{"hello":"Hello world","alphabet":[],"fastField":"This field resolves fast! ⚡️"},"hasNext":true}",
    "timestamp": 1718252802585,
  },
  {
    "content": "{"incremental":[{"items":["a"],"path":["alphabet",0]}],"hasNext":true}",
    "timestamp": 1718252802585,
  },
  {
    "content": "{"incremental":[{"items":["b"],"path":["alphabet",1]}],"hasNext":true}",
    "timestamp": 1718252802585,
  },
  {
    "content": "{"incremental":[{"items":["c"],"path":["alphabet",2]}],"hasNext":true}",
    "timestamp": 1718252802585,
  },
  {
    "content": "{"incremental":[{"items":["d"],"path":["alphabet",3]}],"hasNext":true}",
    "timestamp": 1718252802585,
  },
  {
    "content": "{"incremental":[{"items":["e"],"path":["alphabet",4]}],"hasNext":true}",
    "timestamp": 1718252802585,
  },
  {
    "content": "{"incremental":[{"data":{"slowField":"This field resolves slowly after 5000ms ⏳"},"path":[]}],"hasNext":true}",
    "timestamp": 1718252802585,
  },
  {
    "content": "{"incremental":[{"items":["f"],"path":["alphabet",5]}],"hasNext":true}",
    "timestamp": 1718252802585,
  },
  {
    "content": "{"incremental":[{"items":["g"],"path":["alphabet",6]}],"hasNext":true}",
    "timestamp": 1718252802585,
  },
  {
    "content": "{"incremental":[{"items":["h"],"path":["alphabet",7]}],"hasNext":true}",
    "timestamp": 1718252802585,
  },
  {
    "content": "{"incremental":[{"items":["i"],"path":["alphabet",8]}],"hasNext":true}",
    "timestamp": 1718252802585,
  },
  {
    "content": "{"incremental":[{"items":["j"],"path":["alphabet",9]}],"hasNext":true}",
    "timestamp": 1718252802585,
  },
  {
    "content": "{"incremental":[{"items":["k"],"path":["alphabet",10]}],"hasNext":true}",
    "timestamp": 1718252802585,
  },
  {
    "content": "{"incremental":[{"items":["l"],"path":["alphabet",11]}],"hasNext":true}",
    "timestamp": 1718252802585,
  },
  {
    "content": "{"incremental":[{"items":["m"],"path":["alphabet",12]}],"hasNext":true}",
    "timestamp": 1718252802585,
  },
  {
    "content": "{"incremental":[{"items":["n"],"path":["alphabet",13]}],"hasNext":true}",
    "timestamp": 1718252802585,
  },
  {
    "content": "{"incremental":[{"items":["o"],"path":["alphabet",14]}],"hasNext":true}",
    "timestamp": 1718252802585,
  },
  {
    "content": "{"incremental":[{"items":["p"],"path":["alphabet",15]}],"hasNext":true}",
    "timestamp": 1718252802585,
  },
  {
    "content": "{"incremental":[{"items":["q"],"path":["alphabet",16]}],"hasNext":true}",
    "timestamp": 1718252802585,
  },
  {
    "content": "{"incremental":[{"items":["r"],"path":["alphabet",17]}],"hasNext":true}",
    "timestamp": 1718252802585,
  },
  {
    "content": "{"incremental":[{"items":["s"],"path":["alphabet",18]}],"hasNext":true}",
    "timestamp": 1718252802585,
  },
  {
    "content": "{"incremental":[{"items":["t"],"path":["alphabet",19]}],"hasNext":true}",
    "timestamp": 1718252802585,
  },
  {
    "content": "{"incremental":[{"items":["u"],"path":["alphabet",20]}],"hasNext":true}",
    "timestamp": 1718252802585,
  },
  {
    "content": "{"incremental":[{"items":["v"],"path":["alphabet",21]}],"hasNext":true}",
    "timestamp": 1718252802585,
  },
  {
    "content": "{"incremental":[{"items":["w"],"path":["alphabet",22]}],"hasNext":true}",
    "timestamp": 1718252802585,
  },
  {
    "content": "{"incremental":[{"items":["x"],"path":["alphabet",23]}],"hasNext":true}",
    "timestamp": 1718252802585,
  },
  {
    "content": "{"incremental":[{"items":["y"],"path":["alphabet",24]}],"hasNext":true}",
    "timestamp": 1718252802585,
  },
  {
    "content": "{"incremental":[{"items":["z"],"path":["alphabet",25]}],"hasNext":true}",
    "timestamp": 1718252802585,
  },
  {
    "content": "{"hasNext":false}",
    "timestamp": 1718252802585,
  },
]
`;

exports[`response-builder auto strategy should append the responses when the first response is a JSON object but not patchable 1`] = `
[
  {
    "content": "{"hello":"Hello world","alphabet":[],"fastField":"This field resolves fast! ⚡️"}",
    "timestamp": 1718252802585,
  },
  {
    "content": "{"hi": "Hi world"}",
    "timestamp": 1718252802585,
  },
]
`;

exports[`response-builder auto strategy should concatenate the responses when the first response is not a JSON object 1`] = `
[
  {
    "content": "This is not a JSON object{"incremental":[{"items":["a"],"path":["alphabet",0]}],"hasNext":true}",
    "timestamp": 1718252802585,
  },
]
`;

exports[`response-builder auto strategy should patch the responses when the first response is patchable 1`] = `
[
  {
    "content": "{
  "data": {
    "hello": "Hello world",
    "alphabet": [
      "a"
    ],
    "fastField": "This field resolves fast! ⚡️"
  }
}",
    "timestamp": 1718252802585,
  },
]
`;

exports[`response-builder concatenate strategy should concatenate the responses 1`] = `
[
  {
    "content": "{"data":{"hello":"Hello world","alphabet":[],"fastField":"This field resolves fast! ⚡️"},"hasNext":true}{"incremental":[{"items":["a"],"path":["alphabet",0]}],"hasNext":true}{"incremental":[{"items":["b"],"path":["alphabet",1]}],"hasNext":true}{"incremental":[{"items":["c"],"path":["alphabet",2]}],"hasNext":true}{"incremental":[{"items":["d"],"path":["alphabet",3]}],"hasNext":true}{"incremental":[{"items":["e"],"path":["alphabet",4]}],"hasNext":true}{"incremental":[{"data":{"slowField":"This field resolves slowly after 5000ms ⏳"},"path":[]}],"hasNext":true}{"incremental":[{"items":["f"],"path":["alphabet",5]}],"hasNext":true}{"incremental":[{"items":["g"],"path":["alphabet",6]}],"hasNext":true}{"incremental":[{"items":["h"],"path":["alphabet",7]}],"hasNext":true}{"incremental":[{"items":["i"],"path":["alphabet",8]}],"hasNext":true}{"incremental":[{"items":["j"],"path":["alphabet",9]}],"hasNext":true}{"incremental":[{"items":["k"],"path":["alphabet",10]}],"hasNext":true}{"incremental":[{"items":["l"],"path":["alphabet",11]}],"hasNext":true}{"incremental":[{"items":["m"],"path":["alphabet",12]}],"hasNext":true}{"incremental":[{"items":["n"],"path":["alphabet",13]}],"hasNext":true}{"incremental":[{"items":["o"],"path":["alphabet",14]}],"hasNext":true}{"incremental":[{"items":["p"],"path":["alphabet",15]}],"hasNext":true}{"incremental":[{"items":["q"],"path":["alphabet",16]}],"hasNext":true}{"incremental":[{"items":["r"],"path":["alphabet",17]}],"hasNext":true}{"incremental":[{"items":["s"],"path":["alphabet",18]}],"hasNext":true}{"incremental":[{"items":["t"],"path":["alphabet",19]}],"hasNext":true}{"incremental":[{"items":["u"],"path":["alphabet",20]}],"hasNext":true}{"incremental":[{"items":["v"],"path":["alphabet",21]}],"hasNext":true}{"incremental":[{"items":["w"],"path":["alphabet",22]}],"hasNext":true}{"incremental":[{"items":["x"],"path":["alphabet",23]}],"hasNext":true}{"incremental":[{"items":["y"],"path":["alphabet",24]}],"hasNext":true}{"incremental":[{"items":["z"],"path":["alphabet",25]}],"hasNext":true}{"hasNext":false}",
    "timestamp": 1718252802585,
  },
]
`;

exports[`response-builder patch strategy should gather all errors from all responses 1`] = `
[
  {
    "content": "{
  "data": {
    "hello": "Hello world",
    "alphabet": [
      "a",
      "b"
    ],
    "fastField": "This field resolves fast! ⚡️"
  },
  "errors": [
    {
      "message": "Error 1"
    },
    {
      "message": "Error 2"
    },
    {
      "message": "Incremental error"
    },
    {
      "message": "Error 3"
    }
  ]
}",
    "timestamp": 1718252802585,
  },
]
`;

exports[`response-builder patch strategy should handle incremental payloads in the old format 1`] = `
[
  {
    "content": "{
  "data": {
    "hello": "Hello world",
    "alphabet": [],
    "fastField": "This field resolves fast! ⚡️",
    "inner": {
      "additionalField": "This field is added in the second response"
    }
  }
}",
    "timestamp": 1718252802585,
  },
]
`;

exports[`response-builder patch strategy should patch responses with extensions 1`] = `
[
  {
    "content": "{
  "data": {
    "hello": "Hello world",
    "alphabet": [
      "a",
      "b"
    ],
    "fastField": "This field resolves fast! ⚡️"
  }
}",
    "timestamp": 1718252802585,
  },
]
`;

exports[`response-builder patch strategy should patch the responses 1`] = `
[
  {
    "content": "{
  "data": {
    "hello": "Hello world",
    "alphabet": [
      "a",
      "b",
      "c",
      "d",
      "e",
      "f",
      "g",
      "h",
      "i",
      "j",
      "k",
      "l",
      "m",
      "n",
      "o",
      "p",
      "q",
      "r",
      "s",
      "t",
      "u",
      "v",
      "w",
      "x",
      "y",
      "z"
    ],
    "fastField": "This field resolves fast! ⚡️",
    "slowField": "This field resolves slowly after 5000ms ⏳"
  }
}",
    "timestamp": 1718252802585,
  },
]
`;
